velox/common/base/PeriodicStatsReportDaemon.cpp

```diff
@@ -0,0 +1,272 @@
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeReadBytes, cacheStats.ssdStats->bytesRead);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeWrittenEntries,
+        cacheStats.ssdStats->entriesWritten);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeWrittenBytes,
+        cacheStats.ssdStats->bytesWritten);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeOpenSsdErrors,
+        cacheStats.ssdStats->openFileErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeOpenCheckpointErrors,
+        cacheStats.ssdStats->openCheckpointErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeOpenLogErrors,
+        cacheStats.ssdStats->openLogErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeDeleteCheckpointErrors,
+        cacheStats.ssdStats->deleteCheckpointErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeGrowFileErrors,
+        cacheStats.ssdStats->growFileErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeWriteSsdErrors,
+        cacheStats.ssdStats->writeSsdErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeWriteCheckpointErrors,
+        cacheStats.ssdStats->writeCheckpointErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeReadSsdErrors,
+        cacheStats.ssdStats->readSsdErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCumulativeReadCheckpointErrors,
+        cacheStats.ssdStats->readCheckpointErrors);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCachedEntries, cacheStats.ssdStats->entriesCached);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCachedRegions, cacheStats.ssdStats->regionsCached);
+    RECORD_METRIC_VALUE(
+        kMetricSsdCacheCachedBytes, cacheStats.ssdStats->bytesCached);
+    REPORT_IF_NOT_ZERO(
+        kMetricSsdCacheCheckpointsRead,
+        cacheStats.ssdStats->checkpointsWritten -
+            lastSsdStats_.checkpointsWritten);
+    REPORT_IF_NOT_ZERO(
+        kMetricSsdCacheCheckpointsWritten,
+        cacheStats.ssdStats->checkpointsRead - lastSsdStats_.checkpointsRead);
+  }
+
+  if (auto* cacheTTLController =
+          velox::cache::CacheTTLController::getInstance()) {
+    RECORD_METRIC_VALUE(
+        kMetricCacheMaxAgeSecs,
+        cacheTTLController->getCacheAgeStats().maxAgeSecs);
+
+    RECORD_METRIC_VALUE(
+        kMetricMemoryCacheNumAgedOutEntries,
+        cacheStats.numAgedOut - lastCacheStats_.numAgedOut);
+    RECORD_METRIC_VALUE(
+        kMetricMemoryCacheNumCumulativeAgedOutEntries, cacheStats.numAgedOut);
+
+    if (cacheStats.ssdStats != nullptr) {
+      RECORD_METRIC_VALUE(
+          kMetricSsdCacheCumulativeAgedOutEntries,
+          cacheStats.ssdStats->entriesAgedOut)
+      RECORD_METRIC_VALUE(
+          kMetricSsdCacheCumulativeAgedOutRegions,
+          cacheStats.ssdStats->regionsAgedOut);
+    }
+  }
+
+  lastCacheStats_ = cacheStats;
```

Comment:
Let's just access ssdStats from cacheStats?

Reply:
We can't. ssdStats is a shared_ptr

Reply:
We can't keep a shared_ptr?